### ISSUE TITLE
Ogre 2.0 support

### ISSUE THREAD

**Issue by hhyyrylainen** [Revolutionary-Games/Thrive#306]
The current version of Thrive fails to compile with Ogre 2.0.
I tried to fix this, but it seems that Ogre Ogg Sound needs quite significant changes to work (because it directly messes with Ogre::MovableObjects)
I already made CMake find a matching CEGUI version for Ogre.
Code here: https://github.com/hhyyrylainen/Thrive/tree/Ogre2

At this point it seems that it would be equally hard to get a grasp of how Ogre Ogg internals work as 
it would be to change audio libraries.
I've had my eye on this one for a while now: https://github.com/R4stl1n/cAudio
It seems that it would be pretty easy to create a component/entity that has an Ogre::SceneNode whose movement then controlled audio.

Would it be possible to add that as a dependency? Or at least as a requirement for using newer Ogre versions.

**Comment by jjonj** [comment]
Hrmm yeah that's a tricky situation.
We could certainly add something like cAudio to the build script if necessary. If you think that would be the right choice I could add it, perhaps you would want to do the interfacing with ogre and our component system though, as it would require some deeper knowledge of ogre 2.0

**Comment by hhyyrylainen** [comment]
I added cAudio to the cmake file.
Also updated to work with CEGUI default and started working on a SoundManager.
Branch here: https://github.com/hhyyrylainen/Thrive/tree/306_ogre2_support

**Comment by jjonj** [comment]
Sounds good! :) I'll start looking to adding cAudio!

**Comment by hhyyrylainen** [comment]
Just finished implementing the new audio system. I had to ditch some properties like priority and the default distance, also fading is not in but it could probably be implemented easily.
With GCC the includes fail to compile so I had to create a pull request (to cAudio) to make it work.

The biggest issue right now is probably lua missing createEntity binding. But someone with more luabind experience should look at that (I tried to do AngelScript style binding where a static function mimics a member function but it didn't want to compile...)

**Comment by hhyyrylainen** [comment]
Looks like the lua scripts need to be changed to not use OgreViewportComponent and instead use a new OgreWorkspaceComponent

**Comment by jjonj** [comment]
Alright so you got it under control? :)

**Comment by hhyyrylainen** [comment]
I can probably do this if I don't have to try to get Ogre::SceneManager::createEntity working with luabind as that seems to be the only impossible thing that needs fixing.

I will probably have to create a patch for the asset svn. Or is there a way to create a pull request for that?

**Comment by hhyyrylainen** [comment]
Getting close to starting the game for the first time with Ogre 2. I got the first crash that isn't entirely related to what I'm doing: the game crashes on the second iteration of camera system `for (auto& value : m_impl->m_entities.addedEntities()) {`
```
terminate called after throwing an instance of 'Ogre::InvalidParametersException'
  what():  OGRE EXCEPTION(2:InvalidParametersException): Object already attached to a SceneNode or a Bone in SceneNode::attachObject at /home/hhyyrylainen/Projects/Ogre/ogre/OgreMain/src/OgreSceneNode.cpp (line 122)
```

Just need to figure out why and actually fix the audio and hope that it renders correctly on the first try and I'm done.

**Comment by hhyyrylainen** [comment]
The c++ side is now fixed I just need to rename some Lua functions.
Currently I'm not adding any preprocessor version conditions (does Lua even have those) so the scripts maybe troublesome to add to the main repo because it would force everyone to use Ogre 2 for the scripts to work.

**Comment by jjonj** [comment]
> I will probably have to create a patch for the asset svn. Or is there a way to create a pull request for that?

I think it may just be the easiest if you just have an external download for the necessary asset changes, despite that being a bit clunky.

> Could it be because all the cameras are attached to the same node?

I'll have to have a closer look at your branch to find out! or did you find a fix for this already?

> Currently I'm not adding any preprocessor version conditions (does Lua even have those) so the scripts maybe troublesome to add to the main repo because it would force everyone to use Ogre 2 for the scripts to work.

I'm not sure what you mean here, I assume everyone will be forced to use ogre 2.0 once this branch is merged!

**Comment by hhyyrylainen** [comment]
> I'll have to have a closer look at your branch to find out! or did you find a fix for this already?

It was probably because cameras are by default attached to the root scene node.
>I'm not sure what you mean here, I assume everyone will be forced to use ogre 2.0 once this branch is merged! 

I've been adding the changes inside `#ifdef` blocks so the C++ code works with both Ogre 1.8 and Ogre 2.0 but I don't have enough experience with lua to do the same in lua so the lua scripts are Ogre version specific.

**Comment by jjonj** [comment]
> in lua so the lua scripts are Ogre version specific.

Yeah that would have to happen runtime, which could easily get ugly. I wouldn't require you to support 1.8! Everyone should be able to get 2.0

**Comment by hhyyrylainen** [comment]
Created pull request: https://github.com/Revolutionary-Games/Thrive/pull/309
Everyone will be forced to use Ogre 2.0 (and other requirements like cAudio and CEGUI default branch) mainly because of the Lua files and the new assets.

**Comment by Moopli** [comment]
We're going to need to version-pin CEGUI, since default changes too much to reliably use. 8e6a163 is the commit that merged in @hhyyrylainen's fix, I guess we should all use that until the next version of CEGUI comes out, when we can switch, do the integration work, and stick to it for a while longer.

**Comment by Moopli** [comment]
Compiling Thrive/src/engine/engine.cpp fails on `#include <CEGUI/InputAggregator.h>`, what version of CEGUI has this InputAggregator header? I tried with default, then I tried with 8e6a163, same problem on both.

**Comment by hhyyrylainen** [comment]
@moopli I just checked both commit 8e6a163 and the latest and both have the file `/include/CEGUI/InputAggregator.h`
Perhaps it isn't getting installed or copied to the include folder?

**Comment by Moopli** [comment]
Oh, I forgot to edit it in, I fixed that problem and now I have a double declaration error in two files included by `Thrive/src/engine/engine.cpp`. `typedef SharedPtr<Texture> TexturePtr;` at `/usr/local/include/OGRE/OgreColourValue.h:31`, and `class TexturePtr;` at `/usr/local/include/CEGUI/RendererModules/Ogre/Renderer.h:59`. Mismatch between dependencies? Maybe the installation didn't update all files? Idk.

**Comment by hhyyrylainen** [comment]
@moopli `class TexturePtr` is the old type for Ogre. Ogre 2.0 uses `typedef SharedPtr<Texture> TexturePtr`. The render module shouldn't forward declare the old format if it properly detects that Ogre 2.0 is used.
This code from `/CEGUI/RendererModules/Ogre/Renderer.h`
```
#if (CEGUI_OGRE_VERSION < ((1 << 16) | (9 << 8) | 0))
class TexturePtr;
#else
```
should prevent it from being declared when using Ogre 2.0

**Comment by Moopli** [comment]
Hmm I just rebuilt and installed Ogre2.0 again to be sure (ran cmake and all) and nothing was updated; and then I ran cmake for CEGUI and there's no message about ogre (ie, it says neither found nor missing). I'm building CEGUI again (installed glfw, triggered rebuild of some components), I'll see if Thrive still has issues in a minute.

**Comment by hhyyrylainen** [comment]
When I configure Thrive with cmake I get this message:
`Finding CEGUI version 9999 for Ogre version 2.0.0`

If Ogre 2.0 is properly installed over the older versions (AFAIK you can't have multiple Ogre versions installed) the message should say version 2.0.0

**Comment by Moopli** [comment]
Yup, got the same thing `-- Finding CEGUI version 9999 for Ogre version 2.0.0`but the two lines afterwards are

    -- USERENDERER
    -- USERENDERER

**Comment by hhyyrylainen** [comment]
You could check if CEGUI finds the right Ogre version as `CEGUI_OGRE_VERSION` looks like it might be set by the CEGUI cmake script.
And what also could be wrong are the preprocessor macros for Ogre 2.0.

For me the CEGUI configured file is in `cegui-fork/build/cegui/include/CEGUI/Config.h`
and it has the `CEGUI_OGRE_VERSION_MAJOR` set to 2

If all that is okay then I'm out of ideas.
Perhaps you should ask @jjonj how he got it compiling on windows.

**Comment by Moopli** [comment]
Thanks, it looks like the preprocessor macros in `include/CEGUI/Config.h` were good, so I elected to simply copy them into `/usr/local/include/CEGUI` and now `thrive/engine/engine.cpp` builds. I guess the CEGUI install target isn't properly configured, which fits with my earlier problems (missing CEGUI headers) and @jjonj having to copy the headers somewhere else so Thrive would see them.

**Comment by Moopli** [comment]
(split this out of previous comment)
Ah, and now I fail on a Werror in a cAudio header.
```
/usr/local/include/cAudio/cAudioString.h:103:24: error: ‘const char* cAudio::toWINSTR(const char*)’ defined but not used [-Werror=unused-function]
     static const char* toWINSTR(const char* str)
```

Edit: I tried doing a release build, and I get the same problem.

Edit2: I decided to just add `add_definitions("-Wno-error")` to CmakeLists.txt, and it still Werrors out on me. Very strange.

**Comment by hhyyrylainen** [comment]
@moopli I had the same issue and fixed it by adding this
```
__attribute__((unused)) static void toWINSTRNotUnused(){

    cAudio::toWINSTR(NULL);
}
```
To `src/engine/sound/caudio_include.h`.
You may need to adjust the `#ifdef`  to get it enabled for your setup.

**Comment by Moopli** [comment]
Oh, thanks! I think jjonj commented it out, I put it back with an `#ifndef _WIN32`. Now I'm stuck somewhere else:
`
~/thrive/src/engine/sound/sound_manager.cpp: In constructor ‘thrive::SoundManager::Implementation::Implementation(const string&)’:
~/thrive/src/engine/sound/sound_manager.cpp:37:23: error: deleting object of abstract class type ‘cAudio::IAudioDeviceList’ which has non-virtual destructor will cause undefined behaviour [-Werror=delete-non-virtual-dtor]
         CAUDIO_DELETE devices;
`

**Comment by hhyyrylainen** [comment]
@Moopli What version of cAudio you are using? I did a patch for that error.
https://github.com/R4stl1n/cAudio/pull/33

**Comment by Moopli** [comment]
Oh, I was using [v2.2.0](https://github.com/R4stl1n/cAudio/releases/tag/v2.2.0), I'll use master then.

**Comment by Moopli** [comment]
Perfect, all the sound code compiles now, thanks. Now I'm dealing with some CEGUIWindow problems, and it looks like these are due to your changes sometimes being reverted by jjonj so he can get it to compile on MinGW (from a git blame, anyway).

**Comment by jjonj** [comment]
Note that I've been commenting on my process on https://github.com/Revolutionary-Games/Thrive/pull/309 instead of here. I am also working on a seperate branch!

**Comment by Moopli** [comment]
I know, and I read everything there. I'm currently working off of the ogre2.0 branch, same as you.

**Comment by jjonj** [comment]
Ah alrighty then!

**Comment by Moopli** [comment]
[Here's the paste of my current errors](https://gist.github.com/Moopli/d2f7bea0195ad66140b6)

**Comment by hhyyrylainen** [comment]
@Moopli That line shouldn't get compiled if `CEGUI_USE_NEW` is defined. Which should be defined by thrive cmake file line 179.
Also my version of the file doesn't have anything on those lines where those errors are
https://github.com/hhyyrylainen/Thrive/blob/306_ogre2_support_squashed/src/gui/CEGUIWindow.cpp

**Comment by Moopli** [comment]
My bad, wrong error log. [I updated the gist](https://gist.github.com/Moopli/d2f7bea0195ad66140b6), I was (and still am) getting errors that are difficult to even parse.

**Comment by hhyyrylainen** [comment]
@Moopli That looks like the errors you get when you try to register a CEGUI listener with invalid parameter types.

**Comment by Moopli** [comment]
Right, I think I've found the source of my woes, this innocuous little function:
```
void
CEGUIWindow::registerEventHandler(
    const std::string& eventName,
    const luabind::object& callback
) const {

    // Lambda must return something to avoid an template error.
    auto callbackLambda = [callback](const CEGUI::EventArgs& args) -> int
        {
            luabind::call_function<void>(callback, CEGUIWindow(static_cast<const CEGUI::WindowEventArgs&>(args).window, false));
            return 0;
        };

    m_window->subscribeEvent(eventName, callbackLambda);
}
```

**Comment by Moopli** [comment]
Alright so to be sure that was the problem, I commented out `m_window->subscribeEvent(eventName, callbackLambda);` (for both lambda-using handler-registering methods) and I get:

```
~/thrive/src/gui/CEGUIWindow.cpp: At global scope:
~/thrive/src/gui/CEGUIWindow.cpp:446:1: error: unused parameter ‘eventName’ [-Werror=unused-parameter]
 CEGUIWindow::registerEventHandler(
 ^
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/ThriveLib.dir/src/gui/CEGUIWindow.cpp.o] Error 1
make[1]: *** [CMakeFiles/ThriveLib.dir/all] Error 2
make: *** [all] Error 2
```
Which means I just need to construct the lambda using a `CEGUI::FunctorCopySlot<thrive::CEGUIWindow::registerKeyEventHandler(const luabind::adl::object&) const::__lambda1>::OpVoid {aka void (thrive::CEGUIWindow::registerKeyEventHandler(const luabind::adl::object&) const::__lambda1::*)(const CEGUI::EventArgs&)}`, or whatever CEGUI wants its callbacks to be constructed with.

**Comment by Moopli** [comment]
Finally solved it: Essentially, creating a CEGUI::Event::Subscriber requires the function passed in to return a bool (the path is more complicated, but this is the crux of it). So I made the lambdas return bool and now CEGUIWindow.cpp compiles.

**Comment by Moopli** [comment]
I don't want to push today, as that would require pulling @jjonj's changes, which I'd prefer to leave for another day, so here's the patch:
```
diff --git a/src/gui/CEGUIWindow.cpp b/src/gui/CEGUIWindow.cpp
index 895ca5b..a5c063f 100644
--- a/src/gui/CEGUIWindow.cpp
+++ b/src/gui/CEGUIWindow.cpp
@@ -449,13 +449,13 @@ CEGUIWindow::registerEventHandler(
 ) const {
 
     // Lambda must return something to avoid an template error.
-    auto callbackLambda = [callback](const CEGUI::EventArgs& args) -> int
+    auto callbackLambda = [callback](const CEGUI::EventArgs& args) -> bool
         {
             luabind::call_function<void>(callback, CEGUIWindow(static_cast<const CEGUI::Wind
             return 0;
         };
 
-    m_window->subscribeEvent(eventName, callbackLambda);
+    m_window->subscribeEvent(eventName, CEGUI::Event::Subscriber(callbackLambda));
 }
 
 void
@@ -470,7 +470,7 @@ CEGUIWindow::registerKeyEventHandler(
     const luabind::object& callback
 ) const {
     // Event doesn't exist anymore //
-    auto callbackLambda = [callback](const CEGUI::EventArgs& args) -> int
+    auto callbackLambda = [callback](const CEGUI::EventArgs& args) -> bool
         {
             luabind::call_function<void>(callback, CEGUIWindow(static_cast<
                     const CEGUI::WindowEventArgs&>(args).window, false),
@@ -478,7 +478,7 @@ CEGUIWindow::registerKeyEventHandler(
             return 0;
         };
     
-    m_window->subscribeEvent(CEGUI::Window::EventCharacterKey, callbackLambda);
+    m_window->subscribeEvent(CEGUI::Window::EventCharacterKey, CEGUI::Event::Subscriber(call
 }
 
 void
```

**Comment by Moopli** [comment]
It built successfully, all unit tests passed, but on running of Thrive, Ogre complains of a bunch of missing dynamic libraries, like `./RenderSystem_GL_d.so.2.0.0` and `./OgreOggSound.so.2.0.0`. The first should be solved fine by copying `./RenderSystem_GL.so` for now, but the second surprises me as I thought we switched to cAudio.

**Comment by hhyyrylainen** [comment]
If I remember correctly, thrive uses ogre plugins file to specify the plugins that should be loaded. Removing the ogre ogg plugin from that file should work.

-------- Alkuperäinen viesti --------
Lähettäjä: Filip Burlacu <notifications@github.com> 
Päivämäärä: 07.05.2015  8.19  (GMT+02:00) 
Saaja: Revolutionary-Games/Thrive <Thrive@noreply.github.com> 
Kopio: Henri Hyyryläinen <henri.hyyrylainen@gmail.com> 
Aihe: Re: [Thrive] Ogre 2.0 support (#306) 

It built successfully, all unit tests passed, but on running of Thrive, Ogre complains of a bunch of missing dynamic libraries, like ./RenderSystem_GL_d.so.2.0.0 and ./OgreOggSound.so.2.0.0. The first should be solved fine by copying ./RenderSystem_GL.so for now, but the second surprises me as I thought we switched to cAudio.

**Comment by Moopli** [comment]
Okay, I removed the plugins we aren't using from `plugins.cfg` and Ogre now loads. CEGUI spits out a few errors now about `MousePassThroughEnabled` no longer being a valid property. I figure there'll be a few more of these. I *did* update the assets folder with the zip @hhyyrylainen provided, so that's not the issue.

**Comment by hhyyrylainen** [comment]
CEGUI renamed Mouse to Cursor. So the fix should be to rename `MousePassThroughEnabled` to `CursorPassThroughEnabled` (Or just directly Mouse to Cursor which should fix all related issues and hopefully not break anything)

**Comment by jjonj** [comment]
You should be able to pull my stuff, It's pretty much purely windows scripts stuff and some minor stuff in cmakelist.

**Comment by Moopli** [comment]
I did try pulling your stuff, but cmake failed to configure and I wasn't in the mood to debug your changes, so I reverted. I'll take another look at it today, after working a bit more to get Thrive to run.

**Comment by Moopli** [comment]
```
CEGUI::UnknownObjectException in function 'CEGUI::WindowFactory* CEGUI::WindowFactoryManager::getFactory(const CEGUI::String&) const' (/home/filip/.thrivelibs/cegui/cegui/src/WindowFactoryManager.cpp:186) : A WindowFactory object, an alias, or mapping for 'Thrive/Tooltip' Window objects is not registered with the system.

Have you forgotten to load a scheme using CEGUI::SchemeManager::createFromFile(..)?
```
@hhyyrylainen did you create your assets zip before 5aea2a4?

**Comment by hhyyrylainen** [comment]
@Moopli I probably did, in my fork of thrive there doesn't seem to such a commit:
```
git show 5aea2a4216eae066256a05b3cb2235ef2e6cfa0e
fatal: bad object 5aea2a4216eae066256a05b3cb2235ef2e6cfa0e
```

Should be quite a quick fix just by adding the new Thrive/Tooltip to the scheme file.

**Comment by hhyyrylainen** [comment]
@Moopli Trying to merge my svn changes to the new assets revision is proving to be quite difficult.

**Comment by Moopli** [comment]
Just the standard svn-works-on-a-file-by-file-basis issues, or are there deeper-seated incompatibilities between the workings of CEGUI default and what we've done with tooltips?

You can ignore all the example gui files if you like, those can be replaced wholesale with new examples courtesy of the CEGUI team.

**Comment by hhyyrylainen** [comment]
@Moopli I thought I fixed it for a second, but I just somehow made all files have the exact same contents twice. Working with svn is not very intuitive...

**Comment by hhyyrylainen** [comment]
Well, svn doesn't seem to want to co-operate as it wants to reject all my changes.
I'll just manually copy the tooltip definition to my assets.

**Comment by hhyyrylainen** [comment]
@Moopli I'm unable to start Thrive because Ogre decided that it doesn't like my drivers anymore:
```
./Thrive
terminate called after throwing an instance of 'Ogre::InternalErrorException'
  what():  OGRE EXCEPTION(7:InternalErrorException): Cannot allocate a Hardware query. This video card doesn't support it, sorry. in GLHardwareOcclusionQuery::GLHardwareOcclusionQuery at /home/hhyyrylainen/Projects/Ogre/ogre/RenderSystems/GL/src/OgreGLHardwareOcclusionQuery.cpp (line 64)
Keskeytetty (luotiin core-tiedosto)
```
Perhaps it's time to get rid of that default Ogre settings dialog (which seems to cause the crash.

But I put together a new assets file which should have the new tooltip type and their looknfeels (potentially I missed some other component but those should be it)
https://mega.co.nz/#!RhlFEJAB!xUANo7rdTkwPhOxEgVay49zZsHDlV3ExAwmvpwhV8zM

**Comment by Moopli** [comment]
Mhmm, now it just segfaults without an error message, maybe related to your problem.

**Comment by hhyyrylainen** [comment]
@Moopli Unless you are using mesa drivers, I doubt it. 
At least now I think that all the assets are fine as Ogre and CEGUI usually give an error message when there are missing files

**Comment by jjonj** [comment]
Try running with gdb and see what it says. It might be the same I had which was caused by SSE2 and is fixable.

**Comment by hhyyrylainen** [comment]
I just updated Ogre to latest v2-0 and now it works. Now I just have to fix the errors that merging master caused.

**Comment by hhyyrylainen** [comment]
@Moopli I fixed an error caused by merging master here:
https://github.com/hhyyrylainen/Thrive/tree/306_ogre2
I also had to apply your fix to the CEGUI event listeners making them return bool, so it might conflict if you try to pull it.

**Comment by TheCreator--** [comment]
Was this fixed?

**Comment by Moopli** [comment]
I believe the impeded label was added because this doesn't yet work on linux. jjonj might know.

**Comment by jjonj** [comment]
> I believe the impeded label was added because this doesn't yet work on linux.

I better know since I added the label =P but yeah moopli is correct! I believe my next assignment will be to get a working linux virtual machine so that I can either give moopli a development environment that works if I don't encounter mooplis problem, or help work on the problem myself if the problem does come up!
